nifi-nar-bundles/nifi-elasticsearch-bundle/nifi-elasticsearch-restapi-processors/src/main/java/org/apache/nifi/processors/elasticsearch/PutElasticsearchRecord.java

```diff
@@ -0,0 +1,264 @@
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.nifi.processors.elasticsearch;
+
+import org.apache.nifi.annotation.behavior.InputRequirement;
+import org.apache.nifi.annotation.documentation.CapabilityDescription;
+import org.apache.nifi.annotation.documentation.Tags;
+import org.apache.nifi.annotation.lifecycle.OnScheduled;
+import org.apache.nifi.components.PropertyDescriptor;
+import org.apache.nifi.elasticsearch.ElasticSearchClientService;
+import org.apache.nifi.elasticsearch.ElasticSearchError;
+import org.apache.nifi.elasticsearch.IndexOperationRequest;
+import org.apache.nifi.expression.ExpressionLanguageScope;
+import org.apache.nifi.flowfile.FlowFile;
+import org.apache.nifi.processor.AbstractProcessor;
+import org.apache.nifi.processor.ProcessContext;
+import org.apache.nifi.processor.ProcessSession;
+import org.apache.nifi.processor.Relationship;
+import org.apache.nifi.processor.exception.ProcessException;
+import org.apache.nifi.record.path.FieldValue;
+import org.apache.nifi.record.path.RecordPath;
+import org.apache.nifi.record.path.RecordPathResult;
+import org.apache.nifi.record.path.util.RecordPathCache;
+import org.apache.nifi.record.path.validation.RecordPathValidator;
+import org.apache.nifi.serialization.RecordReader;
+import org.apache.nifi.serialization.RecordReaderFactory;
+import org.apache.nifi.serialization.record.Record;
+import org.apache.nifi.serialization.record.RecordFieldType;
+import org.apache.nifi.serialization.record.util.DataTypeUtils;
+
+import java.io.InputStream;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+
+@InputRequirement(InputRequirement.Requirement.INPUT_REQUIRED)
+@Tags({"json", "elasticsearch", "elasticsearch5", "elasticsearch6", "put", "index", "record"})
+@CapabilityDescription("A record-aware ElasticSearch put processor that uses the official Elastic REST client libraries.")
+public class PutElasticsearchRecord extends AbstractProcessor implements ElasticSearchRestProcessor {
+    static final PropertyDescriptor RECORD_READER = new PropertyDescriptor.Builder()
+        .name("put-es-record-reader")
+        .displayName("Record Reader")
+        .description("The record reader to use for reading incoming records from flowfiles.")
+        .identifiesControllerService(RecordReaderFactory.class)
+        .required(true)
+        .build();
+
+    static final PropertyDescriptor OPERATION_RECORD_PATH = new PropertyDescriptor.Builder()
```

Comment:
If this is specified, does the field remain in the document? What if the user didn't want to have that field included in the document, but needed it in order to specify the operation as a field?

What if the index operation were in an attribute versus a record field? Like if you partitioned records based on whether they were upsert vs delete operations (for example). Perhaps consider something like UpdateRecord's "Replacement Value Strategy" property, which allows you to choose whether the Operation field would be evaluated as a Record Path or a Literal Value. Both can support EL, but the latter would allow the setting of a single operation for all the records in the flow file.

Reply:
The purpose of this capability is to be able to specify bulk operations. In ElasticSearch, you can specify the whole range of CRUD operations in a single bulk operation. That's what this is going for. So you can say "add this, this and this, delete this, upsert that and delete this thing too" in one push.

It has the potential to be quite useful to teams doing a lot of work with ElasticSearch because they will not have to segment changes into different flows, but can create a single flow that is able to take actions to be performed and bulk deliver them to ES through one pipeline.

Reply:
Sorry, early morning and coffee didn't kick in.

Yeah, those values should be removed and they are now. I moved to just purge nulls and empty strings because they're semantically meaningless AFAICT in Lucene and ES since they are truthy-false whether they don't exist, are null or empty. So all gone now if you specify the record path.

Reply:
This still feels a little awkward to me for the case when you want a single operation. I like the idea of being able to specify a RecordPath for the case when you are doing a "bulk" operation but performing multiple types of operations inside it, but if I have a bunch of records I just want to update, I need to add a synthetic field to each with a value of `update`. If that field gets indexed/updated in Elasticsearch, then it's unnecessary for the single-operation case. If instead it gets removed before index/update, then if it wasn't synthetic (such as in a Change Data Capture case where the operation is an existing valid field that needs to be in the doc) then we are removing user data.

Not sure the best way to tackle, but it seems like something you'd want to just pick from, like "use this RecordPath and keep the field in the doc" or "use this value after evaluating EL".  If the incoming data is JSON this might be a moot point since one of the other processors can probably handle the single-operation case, but if it can't or the data isn't JSON, you'd need an UpdateRecord first, to convert the data and/or add a synthetic operation field. What do you think?

Reply:
I'll have to think about it. Might need to reevaluate a few things.